### PR TITLE
Remove `missing_docs` lint on private 2.0 macros

### LINKED ISSUE #57569: missing_docs lint triggers on all declarative macros 2.0
```rust
#![feature(decl_macro)]
#![deny(missing_docs)]
#![allow(unused)]
//! my crate

mod m {
    macro a { () => {} }
    pub(crate) macro b { () => {} }
    pub macro c { () => {} }
    macro_rules! r { () => {} }
    pub fn f() {}
}
```

```
error: missing documentation for macro
 --> m.rs:7:5
  |
7 |     macro a { () => {} }
  |     ^^^^^^^
  |
note: lint level defined here
 --> m.rs:2:9
  |
2 | #![deny(missing_docs)]
  |         ^^^^^^^^^^^^

error: missing documentation for macro
 --> m.rs:8:5
  |
8 |     pub(crate) macro b { () => {} }
  |     ^^^^^^^^^^^^^^^^^^

error: missing documentation for macro
 --> m.rs:9:5
  |
9 |     pub macro c { () => {} }
  |     ^^^^^^^^^^^

```
Note there is no error for `r` or `f`.

### PR DESCRIPTION
**Pull request by inquisitivecrystal**
https://github.com/rust-lang/rust/blob/798baebde1fe77e5a660490ec64e727a5d79970d/compiler/rustc_lint/src/builtin.rs#L573-L584

This code is the source of #57569. The problem is subtle, so let me point it out. This code makes the mistake of assuming that all of the macros in `krate.exported_macros` are exported.

...Yeah. For some historical reason, all `macro` macros are marked as exported, regardless of whether they actually are, which is dreadfully confusing. It would be more accurate to say that `exported_macros` currently contains only macros that have paths. 

This PR renames `exported_macros` to `importable_macros`, since these macros can be imported with `use` while others cannot. It also fixes the code above to no longer lint on private `macro` macros, since the `missing_docs` lint should only appear on exported items.

Fixes #57569.

### CODE REVIEW

**Comment by petrochenkov** [comment]
>since these macros can be imported with `use` while others cannot. 

Other macros can also be imported with `use`.
I suggest avoiding renaming in this PR.
Ideally, `exported_macros` and `non_exported_macro_attrs` should be removed entirely, and macro items should be kept in HIR as regular items, but it requires some work (https://github.com/rust-lang/rust/pull/79396#issuecomment-733534180).

**Comment by inquisitivecrystal** [comment]
> > since these macros can be imported with `use` while others cannot.
> 
> Other macros can also be imported with `use`.
> I suggest avoiding renaming in this PR.
> Ideally, `exported_macros` and `non_exported_macro_attrs` should be removed entirely, and macro items should be kept in HIR as regular items, but it requires some work ([#79396 (comment)](https://github.com/rust-lang/rust/pull/79396#issuecomment-733534180)).

This contains both 2.0 macros and exported macro_rules! macros. Procedural macros are also importable, but at the same time, they're also exported. I'm not entirely sure where they're placed in this hierarchy. Still, it would seem to me that this is strictly more correct and less confusing than the current system? Or am I missing something?

I looked at making macro items work like other items, but that's well above my pay grade.

File: compiler/rustc_lint/src/builtin.rs [hunk 1]
```diff
@@ -570,7 +570,12 @@ impl<'tcx> LateLintPass<'tcx> for MissingDoc {
     fn check_crate(&mut self, cx: &LateContext<'_>, krate: &hir::Crate<'_>) {
         self.check_missing_docs_attrs(cx, hir::CRATE_HIR_ID, krate.item.inner, "the", "crate");
 
-        for macro_def in krate.exported_macros {
+        for macro_def in krate.importable_macros {
+            // Non exported MBE 2.0 macros should be skipped
```
Comment (oli-obk) [inline compiler/rustc_lint/src/builtin.rs:574]: Please also mention the "why" in this comment

**Review by oli-obk — COMMENTED** [review]
> I suggest avoiding renaming in this PR.

I don't have an opinion here, so let's keep the original name and figure out the details on renaming in an issue or zulip thread.

btw, thanks for doing this commit-by-commit, this PR was great to review.

**Comment by inquisitivecrystal** [comment]
> > I suggest avoiding renaming in this PR.
> 
> I don't have an opinion here, so let's keep the original name and figure out the details on renaming in an issue or zulip thread.

Alright! Which would be better, an issue or a Zulip thread? I can do either.
> btw, thanks for doing this commit-by-commit, this PR was great to review.

Thank you!

File: compiler/rustc_lint/src/builtin.rs (hunk 1 above)
Comment (inquisitivecrystal) [reply compiler/rustc_lint/src/builtin.rs]: Done!

**Comment by inquisitivecrystal** [comment]
Alright, the changes are done! I force-pushed to remove the commit that did the rename and adjust the other commits accordingly. My new changes are in b49936c. I made that addition to the comment, as well as changing the code to apply the exported check to `macro_rules` macros as well. There's no reason not to apply it to all macros, and it should also hopefully make my PR compatible with #78855.

**Comment by oli-obk** [comment]
> Alright! Which would be better, an issue or a Zulip thread? I can do either.

for discussion, probably a zulip thread

Let's move this PR on then:

@bors r+

**Pull request merged**